Support for reload_connections when using SSL or basic auth

omrimos opened the pull request:
Added support for reload_connections/reload_on_failure when using basic auth (user/password) and https connections (i.e. when using Shield)

kkirsche commented:
Can you fix the failing tests?

karmi commented:
@kkirsche There are no failing tests -- there's an error at Travis, `Gem::RemoteFetcher::FetchError: Errno::ETIMEDOUT: Connection timed out - connect(2) (https://rubygems.org/gems/slop-3.6.0.gem)`. Please make sure we provide the most valuable feedback first (overall approach, code, etc), and *then* focus on technical details.

karmi commented:
Thanks for the patch, @omrimos! I've got two issues at the moment:

* It seems like your commit merges together a different way of handling setting username/password and changing the regexp for extracting the URLs from Nodes Info.

* There are no tests for this behaviour. A set of Unit tests would be a nice way how to describe the added functionality -- I assume it's based on our offline discussion, and tries to keep the authentication parameters in the host configuration. I'm afraid I can't merge it without tests.

omrimos commented:
* Added tests for all new features:
  * Changed sniffer logic to take http.publish_host (if exists) on reloading instead if plain IP
  * Added the ability to set user/pass (basic auth) and https for the client so that it will remain consistent across connection reloads (after failures as well)

karmi commented:
Hi Omri, sorry for the delayed response! I've changed the commit message to be a bit more descriptive, and did small tweaks here and there. Also, I've rebased it against current master, when a change related to Elasticsearch 2.0 sneaked in. Thank you for the patch -- can you please verify this works in your environment?
